### PR TITLE
fix(meeting-plugin):  send missing media quality metrics

### PR DESCRIPTION
**Pull request by arun3528**
TBD screen shots for logs and MQE

https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-186042

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js [hunk 1]
```diff
@@ -239,6 +252,149 @@ export default class StatsAnalyzer extends EventsScope {
...
+   *  @param {String} result
+   * @param {boolean} type
+   * @returns {void}
+   */
+  parseAudioSource(result, type) {
+    if (!result) {
+      return;
+    }
+
+    if (type === 'audio') {
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js:266]: Would like to see all the media types using constants.

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 2]
```diff
@@ -966,9 +966,12 @@ export default class Meeting extends StatelessWebexPlugin {
    * @memberof Meeting
    */
   setUpLocusMediaSharesListener() {
+    // Will get triggred on local and remote share
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:969]:
```suggestion
    // Will get triggered on local and remote share
```

File: packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js [hunk 3]
```diff
@@ -58,6 +64,12 @@ SelfUtils.getSelves = (oldSelf, newSelf, deviceId) => {
 
   updates.isMediaInactive = SelfUtils.isMediaInactive(previous, current);
 
+  console.log('MEDIA STATUS ', current?.currentMediaStatus);
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/locus-info/selfUtils.js:67]: Need to use logger here

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 4]
```diff
@@ -974,9 +974,12 @@ export default class Meeting extends StatelessWebexPlugin {
    * @memberof Meeting
    */
   setUpLocusMediaSharesListener() {
+    // Will get triggered on local and remote share
     this.locusInfo.on(EVENTS.LOCUS_INFO_UPDATE_MEDIA_SHARES, (payload) => {
       const {contentId, disposition} = payload.current;
 
+      if (this.statsAnalyzer) this.statsAnalyzer.updateShareStatus(disposition === FLOOR_ACTION.GRANTED);
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:981]: Please space this out so eslint doesn't complain or use the optional chaining method:

```suggestion
      this.statsAnalyzer?.updateShareStatus(disposition === FLOOR_ACTION.GRANTED);
```

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 5]
```diff
@@ -1208,6 +1211,26 @@ export default class Meeting extends StatelessWebexPlugin {
       );
       this.reconnect();
     });
+
+    // There is two stats for mute one is the actual media being sent or received
+    // The second on is if the audio is muted, we need to tell the statsAnalyzer when
+    // the audio is muted or the user is not willing to send media
+    this.locusInfo.on(LOCUSINFO.EVENTS.MEDIA_STATUS_CHANGE, (status) => {
+      if (this.statsAnalyzer) {
+        this.statsAnalyzer.updateMediaStatus({
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:1220]:
```suggestion
        this.statsAnalyzer?.updateMediaStatus({
```

File: packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js [hunk 6]
```diff
@@ -239,6 +264,159 @@ export default class StatsAnalyzer extends EventsScope {
     });
   }
 
+  /**
+   * parse the audio and video source
+   *  @param {String} result
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js:269]:
```suggestion
   * @param {String} result
```

File: packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js [hunk 7]
```diff
@@ -413,16 +601,22 @@ export default class StatsAnalyzer extends EventsScope {
         LoggerProxy.logger.info('StatsAnalyzer:index#processInboundRTPResult --> Packets getting lost from the receiver ', this.statsResults[mediaType][sendrecvType].currentPacketLossRatio);
       }
 
-      this.statsResults[mediaType][sendrecvType].totalPacketsLost = result.packetsLost;
+      // TODO: check the packet loss value is negative values here
+
+      if (result.packetsLost) {
+        this.statsResults[mediaType][sendrecvType].totalPacketsLost = result.packetsLost > 0 ? result.packetsLost : -result.packetsLost;
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js:607]: We are just assuming a negative packet loss needs to be switched to positive?

**Review by adamweeks — CHANGES_REQUESTED** [review]

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 4 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:981]: @adamweeks i got a lint error when i do that , Keeping it like that

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 5 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:1220]: same error , We cannot call that over a function

File: packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js (hunk 7 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js:607]: I have seen in chrome the packet loss was negative values, So need to change that

**Comment by arun3528** [comment]
@adamweeks  ready for review

File: packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js (hunk 7 above)
Comment (adamweeks) [reply packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js:607]: Should we change a negative value to 0 instead? Changing a `-100` to `100` seems like a bad solution unless we know for a fact it is a browser bug that is reporting the value incorrectly.

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 4 above)
Comment (adamweeks) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:981]: I've noticed the eslint error as well when trying to use the optional chaining. We'll need to update the eslint config. In the meantime, you can either disable the eslint warning or move this to a multi line `if`:

```suggestion
      if (this.statsAnalyzer) {
        this.statsAnalyzer.updateShareStatus(disposition === FLOOR_ACTION.GRANTED);
      }
```

File: packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js [hunk 8]
```diff
@@ -107,6 +120,9 @@ export default class StatsAnalyzer extends EventsScope {
       isShareStream: true
     });
 
+    mqaData.intervals[0].intervalMetadata.peerReflexiveIP = this.statsResults.connectionType.local.ipAddress[0];
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js:123]: Do we need to check if the `intervals` array exists and has data?

File: packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js [hunk 9]
```diff
@@ -239,6 +264,159 @@ export default class StatsAnalyzer extends EventsScope {
     });
   }
 
+  /**
+   * parse the audio and video source
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js:268]:
```suggestion
   * parse the audio
```

File: packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js [hunk 10]
```diff
@@ -281,6 +460,20 @@ export default class StatsAnalyzer extends EventsScope {
...
+    this.statsResults[STATS.VIDEO_CORRELATE].direction = this.peerConnection.videoTransceiver.currentDirection;
+    this.statsResults[STATS.SHARE_CORRELATE].direction = this.peerConnection.shareTransceiver.currentDirection;
+
+    // Process Stats results every 5 seconds
+
+    this.compareLastStatsResult();
+
+    // Save the last results to compare with the current
+    // DO Deep copy, for some reason it takes the reference all the time rather then old value set
+    this.lastStatsResults = JSON.parse(JSON.stringify(this.statsResults));
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js:475]: Can you use the `cloneDeep` lodash function instead as it is used elsewhere in the sdk?

```js
import {cloneDeed} from 'lodash';
```

File: packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js [hunk 11]
```diff
@@ -395,7 +583,7 @@ export default class StatsAnalyzer extends EventsScope {
       this.statsResults[mediaType].bytesReceived = kilobytes.toFixed(1);
 
       this.statsResults[mediaType][sendrecvType].pliCount = result.pliCount - this.statsResults.internal[mediaType][sendrecvType].pliCount;
-      this.statsResults[mediaType][sendrecvType].currentPacketsLost = result.packetsLost - this.statsResults.internal[mediaType][sendrecvType].packetsLost;
+      this.statsResults[mediaType][sendrecvType].currentPacketsLost = this.statsResults.internal[mediaType][sendrecvType].packetsLost - result.packetsLost;
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js:586]: Why are you swapping these? Isn't `result.packetsLost` the total packets?

File: packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js (hunk 7 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js:607]: I was seeing packet loss as -100 meaning around 100 packets were lost

File: packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js (hunk 8 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js:123]: There is a global file which has a meta data for the intervals array

File: packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js (hunk 10 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js:475]: will give a try i think i tried clone before and it was taking it as a reference

File: packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js [hunk 12]
```diff
@@ -239,6 +265,159 @@ export default class StatsAnalyzer extends EventsScope {
...
+      let mediaType = STATS.AUDIO_CORRELATE;
+      let currentStats = null;
+      let previousStats = null;
+
+      if (this.meetingMediaStatus.expected.sendAudio) {
+        currentStats = this.statsResults[mediaType].send;
+        previousStats = this.lastStatsResults[mediaType].send;
+
+        if (currentStats.totalPacketsSent === previousStats.totalPacketsSent || currentStats.totalPacketsSent === 0) {
+          LoggerProxy.logger.log(`StatsAnalyzer:index#compareLastStatsResult --> No ${mediaType} RTP packets sent`);
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/statsAnalyzer/index.js:304]: Do all of these need to be at `log` level? `info` or `debug` might reduce the noise, but if you think the end user really needs to know about them, we can leave at log.

**Review by adamweeks — APPROVED** [review]
Approved, just one minor request about the log level, but not a merging blocker. Great job!

**Pull request merged**